### PR TITLE
fix(codex): run Windows shell hooks non-interactively

### LINKED ISSUE #1136: Codex: SessionStart and UserPromptSubmit open persistent Git Bash windows on Windows
### 📝 Bug Description

Al utilizar Engram con Codex dentro de Orca en Windows, los hooks `SessionStart` y `UserPromptSubmit` abren ventanas visibles de Git Bash / Mintty que permanecen abiertas incluso después de que Codex considere terminado el hook.

El problema aparece porque estos dos hooks no tienen actualmente un `commandWindows` específico y se ejecuta directamente la ruta `.sh`.

En Windows, esa ruta termina resolviéndose mediante la asociación de archivos `.sh` de Git for Windows, lo que abre Git Bash / Mintty de forma interactiva.

Durante sesiones largas pueden acumularse varias ventanas y procesos `git-bash.exe`, `mintty.exe` y `bash.exe`.

He podido reproducir el problema varias veces, identificar el árbol de procesos y comprobar que una invocación explícita de Bash no interactivo elimina las ventanas y los procesos residuales.

Actualmente utilizo un workaround temporal que desactiva únicamente `SessionStart` y `UserPromptSubmit` en Windows, manteniendo Engram MCP disponible manualmente.

### 🔄 Steps to Reproduce

1. Instalar/activar Engram para Codex en Windows.
2. Abrir una sesión de Codex dentro de Orca.
3. Iniciar una sesión o enviar un prompt.
4. Engram ejecuta `SessionStart` o `UserPromptSubmit`.
5. Aparece una ventana visible de Git Bash / Mintty.
6. La ventana permanece abierta después de que Codex considere terminado el hook.

El árbol de procesos observado es aproximadamente:

    Codex / PowerShell
    → asociación de archivos .sh de Windows
    → git-bash.exe --no-cd
    → mintty.exe
    → bash.exe --login -i <hook.sh>

En la máquina afectada, la asociación de `.sh` es:

    .sh=sh_auto_file

    sh_auto_file="C:\Program Files\Git\git-bash.exe" --no-cd "%L" %*

Los hooks afectados leen el payload completo desde stdin mediante una operación equivalente a:

    INPUT=$(cat)

Al ejecutarse dentro de Git Bash / Mintty interactivo, `cat` queda esperando EOF y el proceso Bash permanece vivo.

### ✅ Expected Behavior

En Windows, `SessionStart` y `UserPromptSubmit` deberían:

- ejecutarse de forma no interactiva;
- consumir correctamente el payload enviado por stdin;
- devolver el stdout esperado a Codex;
- esperar al proceso real del hook hasta su finalización;
- no abrir una ventana visible;
- no depender de la asociación `.sh` de Windows;
- no iniciar Mintty;
- no dejar procesos `git-bash.exe`, `mintty.exe` o `bash.exe` residuales;
- conservar correctamente la integración de Engram con sesión, proyecto y memoria.

### ❌ Actual Behavior

Codex/PowerShell ejecuta la ruta `.sh` y Windows la abre mediante la asociación de Git Bash.

El resultado observado es:

    PowerShell
    → git-bash.exe
    → mintty.exe
    → bash.exe --login -i <script>

El proceso Bash queda bloqueado esperando stdin/EOF y la ventana permanece abierta.

Mientras tanto, Codex puede registrar el launcher como terminado correctamente, por ejemplo:

    exit code 0
    no output

aunque el proceso real del hook siga ejecutándose.

Se observaron procesos residuales tanto para:

    session-start.sh

como para:

    user-prompt-submit.sh

No se trata únicamente de un `conhost.exe` huérfano: el propio `bash.exe` permanece vivo.

También se probó una ejecución explícita mediante Bash no interactivo. En ese caso:

- exit code 0;
- no apareció Mintty;
- no apareció una ventana visible;
- no quedaron nuevos procesos `git-bash.exe`, `mintty.exe` o `bash.exe`.

Esto confirma que el problema de las ventanas persistentes está relacionado con ejecutar el `.sh` mediante la asociación de archivos de Windows.

### Operating System

Windows

### Engram Version

Engram CLI v1.20.0

### Agent / Client

Other

### 📋 Relevant Logs

```shell
Observed process chain:

PowerShell / Codex
→ git-bash.exe --no-cd
→ mintty.exe
→ bash.exe --login -i ...\session-start.sh

Equivalent residual process chain also observed for:

...\user-prompt-submit.sh

Windows .sh association:

.sh=sh_auto_file
sh_auto_file="C:\Program Files\Git\git-bash.exe" --no-cd "%L" %*

Codex reports the launcher as completed:

exit code 0
no output

but git-bash.exe / mintty.exe / bash.exe remain alive.

Controlled non-interactive Bash execution:

- exit code 0
- no Mintty window
- no new git-bash.exe residual process
- no new mintty.exe residual process
- no new bash.exe residual process
```

### 💡 Additional Context

Codex CLI 0.154.0 running inside Orca 1.4.199

Los hooks afectados son:

- `SessionStart`
- `UserPromptSubmit`

Archivos relacionados:

    plugin/codex/hooks/hooks.json
    plugin/codex/scripts/session-start.sh
    plugin/codex/scripts/user-prompt-submit.sh

Otros hooks de Engram que ya disponen de `commandWindows`, como `SessionEnd` / `SubagentStop`, no presentaron este mismo comportamiento durante las pruebas.

Inicialmente una prueba con Bash no interactivo también encontró que `jq` no estaba disponible. Se instaló posteriormente `jq 1.8.2`, por lo que `jq` ya no es el blocker principal.

Después de instalar `jq` se ejecutaron de forma controlada:

- 1 × `SessionStart`
- 3 × `UserPromptSubmit`

mediante Git Bash no interactivo.

Los procesos terminaron con exit code 0 y no dejaron ventanas ni procesos residuales nuevos.

Sin embargo, `SessionStart` mostró otro problema relacionado con `engram instance-id`: el entorno probado no podía resolver correctamente ese comando. El binario local identificado era Engram CLI v1.20.0, mientras que el plugin Codex instalado era 0.1.4. Esto puede ser un problema de compatibilidad separado o relacionado que los mantenedores quizá quieran revisar. :contentReference[oaicite:2]{index=2}

Como workaround temporal local he desactivado únicamente `SessionStart` y `UserPromptSubmit` en Windows usando `commandWindows` con un comando que termina inmediatamente.

Con ese workaround:

- no aparecen nuevas ventanas Git Bash / Mintty;
- no aparecen nuevos procesos Bash residuales;
- Engram MCP sigue accesible manualmente;
- las herramientas de memoria siguen funcionando.

El workaround implica perder temporalmente la automatización de esos dos hooks y no lo considero una solución definitiva.

Creo que la dirección de solución debería ser evitar que Windows ejecute los `.sh` afectados mediante `sh_auto_file`, y proporcionar una ruta Windows explícita (`commandWindows` o equivalente) que preserve correctamente stdin, stdout, timeouts, fail-open y la integración con sesión/proyecto/memoria.

### PR DESCRIPTION
**Pull request by dnlrsls**
## 🔗 Linked Issue

Closes #1136

---

## 🏷️ PR Type

- [x] `type:bug` — Bug fix
- [ ] `type:feature` — New feature
- [ ] `type:question` — Question requiring tracked work
- [ ] `type:docs` — Documentation only
- [ ] `type:refactor` — Code refactoring (no behavior change)
- [ ] `type:chore` — Maintenance, dependencies, tooling
- [ ] `type:breaking-change` — Breaking change

---

## 📝 Summary

- Add explicit non-interactive Windows dispatch for Codex `SessionStart`, post-compaction, and `UserPromptSubmit` shell hooks.
- Reuse the existing Bash hook logic through Git-for-Windows `bash.exe` without invoking `git-bash.exe` or Mintty.
- Preserve stdin EOF, Unicode stdout/stderr, child exit status, timeout behavior, and fail-open semantics across Windows console code-page changes.

## 📂 Changes

| File | Change |
|------|--------|
| `plugin/codex/hooks/hooks.json` | Route all three remaining `.sh` hooks through a Windows-specific dispatcher. |
| `plugin/codex/scripts/run-bash-hook.ps1` | Discover Git-for-Windows Bash, validate the approved script path, and relay hook stdio without opening Mintty. |
| `plugin/codex_windows_hook_dispatcher_test.go` | Cover manifest mappings, paths with spaces, EOF, Unicode streams, wait/exit behavior, fail-open cases, and code-page/BOM regression. |
| `plugin/codex/.codex-plugin/plugin.json` | Bump the Codex plugin version to `0.1.5`. |
| `plugin/codex_windows_session_end_test.go` | Update the expected plugin version. |

## 🧪 Test Plan

- [ ] Unit tests pass locally: `go test ./...`
- [ ] E2E tests pass locally: `go test -tags e2e ./internal/server/...`
- [ ] Lint passes locally: `make lint`
- [x] Manually tested the affected functionality

Focused verification completed:

- [x] `go test ./plugin -run '^TestCodexWindowsBashHookDispatcher' -count=1 -v`
- [x] `go test ./plugin -run '^TestCodexWindows(BashHookDispatcherRuntime|SubagentStopAdapter)$' -count=3 -v`
- [x] `go test ./plugin -run '^TestCodexWindows' -count=1`
- [x] PowerShell syntax parse for `run-bash-hook.ps1`
- [x] `git diff --check`
- [x] Real Orca/Codex smoke test with locally installed Engram plugin `0.1.5`: startup `SessionStart`, `UserPromptSubmit`, and compact `SessionStart` completed with expected context and no Mintty or residual hook Bash process.

Local broad-suite limitations:

- `go test ./...` and the E2E command fail on this Windows host because of existing project-resolution/Git-Bash-path and unrelated plugin-test failures. Relevant failures were reproduced against an exported unmodified `HEAD`; no candidate-caused broad-suite failure was found.
- `make lint` was unavailable because `make` is not installed.
- `plugin/pi` tests were not run because dependencies are not installed. CI is the authoritative clean-environment gate for this draft.

---

## 🤖 Automated Checks

These run automatically and **all must pass** before merge:

| Check | What it verifies | Status |
|-------|-----------------|--------|
| **Check Issue Reference** | PR body contains `Closes #N` / `Fixes #N` / `Resolves #N` | ⏳ |
| **Check Issue Has status:approved** | Linked issue has `status:approved` label | ⏳ |
| **Check PR Has type:* Label** | Canonical labels, applicability, and cardinality | ⏳ |
| **Check PR Has No Transient Artifacts** | PR files comply with the Transient Artifact Policy | ⏳ |
| **Unit Tests** | `go test ./...` passes | ⏳ |
| **E2E Tests** | `go test -tags e2e ./internal/server/...` passes | ⏳ |
| **Plugin Tests** | `npm test` passes in `plugin/pi` | ⏳ |
| **Lint** | golangci-lint reports no new findings | ⏳ |

---

## ✅ Contributor Checklist

- [x] I linked an approved issue above (`Closes #1136`)
- [x] I added exactly **one** `type:*` label to this PR
- [ ] I ran unit tests locally: `go test ./...`
- [ ] I ran e2e tests locally: `go test -tags e2e ./internal/server/...`
- [ ] I ran lint locally: `make lint`
- [x] Docs reviewed; no user-facing workflow documentation change was required
- [x] Commits follow conventional commits format
- [x] No `Co-Authored-By` trailers in commits
- [x] I checked every changed path against the Transient Artifact Policy

---

## 💬 Notes for Reviewers

- `size:exception` was explicitly approved for this 442-line cohesive work unit. The dispatcher, manifest mappings, and behavior-focused Windows regression coverage do not have an honest independently useful split below the 400-line budget.
- The candidate completed a four-lens native review (risk, resilience, readability, and reliability) with approval; remaining advisories were informational only.
- No Bash hook business logic or output contract was changed. `jq` availability and the separately reported `engram instance-id` compatibility concern remain out of scope.

## Summary by CodeRabbit

- **New Features**
  - Added Windows support for additional Codex plugin hooks, enabling Bash scripts to run through PowerShell.
  - Improved Unicode input and output handling for Windows hook execution.

- **Bug Fixes**
  - Added safeguards for unsupported hooks and unavailable Git Bash installations, allowing hooks to exit cleanly.

- **Chores**
  - Updated the Codex plugin version to 0.1.5.

### CODE REVIEW

File: plugin/codex_windows_hook_dispatcher_test.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	}
+	if testing.Short() {
+		t.Skip("executes the console-encoding regression through Git for Windows Bash")
+	}
+	if _, ok := codexGitForWindowsBash(); !ok {
+		t.Skip("Git for Windows Bash unavailable: git.exe does not resolve to an installation with bin/bash.exe or usr/bin/bash.exe")
+	}
+
+	originalCodePage := codexWindowsConsoleCodePage(t)
+	t.Cleanup(func() { codexSetWindowsConsoleCodePage(t, originalCodePage) })
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply plugin/codex_windows_hook_dispatcher_test.go]: Fixed in eecb4cc. The test now captures input and output console code pages separately, sets them to different values in the regression, restores the original pair in cleanup, and bounds the PowerShell helpers with an 8-second context.

File: plugin/codex_windows_hook_dispatcher_test.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	command := strings.ReplaceAll(codexBashHookWindowsCommand(t, root, `"${PLUGIN_ROOT}/scripts/session-start.sh"`), "${PLUGIN_ROOT}", pluginRoot)
+	stdout, stderr, code := runCodexWindowsManifestCommand(t, command, input, "")
+	// The fixture sleeps before exiting, so its exit code proves the dispatcher waited for Bash.
+	if code != 23 {
+		t.Fatalf("exit=%d stdout=%q stderr=%q, want child exit 23", code, stdout, stderr)
+	}
+	if stdout != wantWorkingDir.String()+wantStdout || stderr != wantStderr {
+		t.Fatalf("stdout=%q stderr=%q, want current directory and Unicode hook output preserved", stdout, stderr)
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply plugin/codex_windows_hook_dispatcher_test.go:143]: Not changing this. Hook timeout and Windows process-tree cleanup are owned by the Codex host runner, which places the command tree in a Windows JobObject and terminates it on timeout. This repository helper launches cmd.exe directly and cannot honestly exercise that host contract; adding a surrogate would test a different runner and duplicate external behavior. The dispatcher intentionally waits for Bash and remains inside the host-owned process tree.

File: plugin/codex/scripts/run-bash-hook.ps1 [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    $scriptPath = [System.IO.Path]::GetFullPath($HookScript)
+    $expectedPath = [System.IO.Path]::GetFullPath((Join-Path $PSScriptRoot $scriptName))
+    if (-not [string]::Equals($scriptPath, $expectedPath, [System.StringComparison]::OrdinalIgnoreCase) -or
+        -not [System.IO.File]::Exists($scriptPath)) {
+        exit 0
+    }
+
+    $git = Get-Command -Name 'git.exe' -CommandType Application -ErrorAction Stop | Select-Object -First 1
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply plugin/codex/scripts/run-bash-hook.ps1:24]: Not changing this. run-bash-hook.ps1 contains Windows transport and process-launch glue only: allowlist validation, Git-for-Windows Bash discovery, raw stdio relay, and child completion. Engram business logic remains in the existing hook scripts and core. Adding a new Go CLI/API solely to launch Bash would introduce another runtime dependency and boundary without reducing complexity.
